api/grpc/mpi/v1/command.proto

```diff
@@ -0,0 +1,387 @@
+}
+
+// Report on the health of a particular instance
+message InstanceHealth {
+    // Health status enum
+    enum InstanceHealthStatus {
+        // Unspecified status
+        INSTANCE_HEALTH_STATUS_UNSPECIFIED = 0;
+        // Healthy status
+        INSTANCE_HEALTH_STATUS_HEALTHY = 1;
+        // Unhealthy status
+        INSTANCE_HEALTH_STATUS_UNHEALTHY = 2;
+        // Degraded status
+        INSTANCE_HEALTH_STATUS_DEGRADED = 3;
+    }
+    string instance_id = 1;
+    // Health status
+    InstanceHealthStatus instance_health_status = 2;
+    // Provides a human readable context around why a health status is a particular state
+    string description = 3;
+}
+
+// Metainformation relating to the reported instance
+message InstanceMeta {
+    // the identifier associated with the instance
+    string instance_id = 1;
+    // the types of instances possible
+    enum InstanceType {
+        // Unspecified instance type
+        INSTANCE_TYPE_UNSPECIFIED = 0;
+        // NGINX Agent
+        INSTANCE_TYPE_AGENT = 1;
+        // NGINX
+        INSTANCE_TYPE_NGINX = 2;
+        // NGINX Plus
+        INSTANCE_TYPE_NGINX_PLUS = 3;
+        // NGINX Unit
+        INSTANCE_TYPE_UNIT = 4;
+    }
+    // the types of instances possible
+    InstanceType instance_type = 2;
+    // the version of the instance
+    string version = 3;
+}
+
+// Health report of a set of instances
+message UpdateDataPlaneHealthRequest {
+    // Meta-information associated with a message
+    mpi.v1.MessageMeta message_meta = 1;
+    // Health report of a set of instances
+    repeated InstanceHealth instance_healths = 2;
+}
+
+// Response to a UpdateDataPlaneHealthRequest - intentionally empty
+message UpdateDataPlaneHealthResponse {}
+
+// Reports the status of an associated command. This may be in response to a ManagementPlaneRequest
+message DataPlaneResponse {
+    // Meta-information associated with a message
+    mpi.v1.MessageMeta message_meta = 1;
+    // The command response with the associated request
+    mpi.v1.CommandResponse command_response = 2;
+}
+
+// A Management Plane request for information, triggers an associated rpc on the DataPlane
+message ManagementPlaneRequest {
+    // Meta-information associated with a message
+    mpi.v1.MessageMeta message_meta = 1;
+    oneof request {
+        // triggers a DataPlaneStatus rpc
+        StatusRequest status_request = 2;              
+        // triggers a DataPlaneHealth rpc
+        HealthRequest health_request = 3;               
+        // triggers a rpc GetFile(FileRequest) for overview list, 
+        // if overview is missing, triggers a rpc Overview(ConfigVersion) first
+        ConfigApplyRequest config_apply_request = 5;
+        // triggers a series of rpc SendFile(File) for that instances
+        ConfigUploadRequest config_upload_request = 6;
+        // triggers a reconciliation of  with a command_response for a particular action
+        ConfigSyncRequest config_sync_request = 7; 
```

Comment:
Why do we need sync if we have apply and upload?

Reply:
if the Agent has been offline for awhile or has local changes not shared with the Management Plane a sync is required no?

Reply:
The workflow currently specifies that the Agent should upload a FileOverview after being offline, and then MP can send a ConfigUploadRequest if there's anything missing. I'd say we don't need Sync.